File: src/main/java/art/arcane/racbot/service/EmbedFactory.java
package art.arcane.racbot.service;

import art.arcane.racbot.config.BotConfig;
import art.arcane.racbot.util.ColorParser;
import java.time.Instant;
import net.dv8tion.jda.api.EmbedBuilder;

public class EmbedFactory {
  private final BotConfig config;

  public EmbedFactory(BotConfig config) {
    this.config = config;
  }

  public EmbedBuilder base(String title) {
    return new EmbedBuilder()
        .setTitle(title)
        .setColor(ColorParser.parseHex(config.branding().primaryColorHex()))
        .setFooter(config.branding().companyName(), config.branding().avatarUrl())
        .setTimestamp(Instant.now());
  }

  public EmbedBuilder info(String title, String description) {
    return base(title).setDescription(description);
  }

  public EmbedBuilder error(String title, String description) {
    return base(title).setDescription(":warning: " + description);
  }
}

File: lib/src/service/embed_factory.dart
import 'package:nyxx/nyxx.dart';
import 'package:racbot_nyxx/src/config/bot_config.dart';
import 'package:racbot_nyxx/src/util/color_parser.dart';

class EmbedFactory {
  final BotConfig config;

  const EmbedFactory({required this.config});

  EmbedBuilder base({required String title}) {
    Uri? avatarUri = Uri.tryParse(config.branding.avatarUrl);
    EmbedFooterBuilder footer = EmbedFooterBuilder(
      text: config.branding.companyName,
      iconUrl: avatarUri,
    );

    return EmbedBuilder(
      title: title,
      color: ColorParser.parseHex(config.branding.primaryColorHex),
      footer: footer,
      timestamp: DateTime.now().toUtc(),
    );
  }

  EmbedBuilder info({required String title, required String description}) =>
      base(title: title)..description = description;

  EmbedBuilder error({required String title, required String description}) =>
      base(title: title)..description = ':warning: $description';
}
